packages/block-library/src/heading/autogenerate-anchors.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * External dependencies
+ */
+import { deburr, trim } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+
+/**
+ * Runs a callback over all blocks, including nested blocks.
+ *
+ * @param {Object[]} blocks   The blocks.
+ * @param {Function} callback The callback.
+ *
+ * @return {void}
+ */
+const recurseOverBlocks = ( blocks, callback ) => {
+	for ( const block of blocks ) {
+		// eslint-disable-next-line callback-return
+		callback( block );
+		if ( block.innerBlocks ) {
+			recurseOverBlocks( block.innerBlocks, callback );
+		}
+	}
+};
+
+/**
+ * Returns the text without markup.
+ *
+ * @param {string} text The text.
+ *
+ * @return {string} The text without markup.
+ */
+const getTextWithoutMarkup = ( text ) => {
+	const dummyElement = document.createElement( 'div' );
+	dummyElement.innerHTML = text;
+	return dummyElement.innerText;
+};
+
+/**
+ * Get all heading anchors.
+ *
+ * @param {Object} blockList An object containing all blocks.
+ * @param {string} excludeId A block ID we want to exclude.
+ *
+ * @return {string[]} Return an array of anchors.
+ */
+const getAllHeadingAnchors = ( blockList, excludeId ) => {
+	const anchors = [];
+
+	recurseOverBlocks( blockList, ( block ) => {
+		if (
+			block.name === 'core/heading' &&
+			( ! excludeId || block.clientId !== excludeId ) &&
+			block.attributes.anchor
+		) {
+			anchors.push( block.attributes.anchor );
+		}
+	} );
+
+	return anchors;
+};
+
+/**
+ * Get the slug from the content.
+ *
+ * @param {string} content The block content.
+ *
+ * @return {string} Returns the slug.
+ */
+const getSlug = ( content ) => {
+	// Get the slug.
+	return trim(
+		deburr( getTextWithoutMarkup( content ) )
+			.replace( /[^\p{L}\p{N}]+/gu, '-' )
+			.toLowerCase(),
+		'-'
+	);
+};
+
+/**
+ * Generate the anchor for a heading.
+ *
+ * @param {string}   content           The block content.
+ * @param {string[]} allHeadingAnchors An array containing all headings anchors.
+ *
+ * @return {string|null} Return the heading anchor.
+ */
+export const generateAnchor = ( content, allHeadingAnchors ) => {
+	const slug = getSlug( content );
+	// If slug is empty, then return null.
+	// Returning null instead of an empty string allows us to check again when the content changes.
+	if ( '' === slug ) {
+		return null;
+	}
+
+	let anchor = slug;
+	let i = 0;
+
+	// If the anchor already exists in another heading, append -i.
+	while ( allHeadingAnchors.includes( anchor ) ) {
+		i += 1;
+		anchor = slug + '-' + i;
+	}
+
+	return anchor;
+};
```

Comment:
Let's say I save a page with three headings that map to the same anchor. Then here we get `plants`, `plants-1`, `plants-2`. Let's say I remove the initial paragraph from that page, and add a new one that also starts with `plants`. I think this code would assign it the freshly freed anchor `plants`, which means that all the links to that anchor would now lead to a wrong heading.

I would love to make these headings persistent in some way. A naive approach would involve always `+1`-ing the largest related number, so even if `plants` is free and we only have `plants-1` and `plants-2`, we would still assign `plants-3`.

The problem with that approach is that it is possible to remove just `plants-2` or even both `plants-1` and `plants-2` so now there is no information we could use to avoid assigning new headings with the same anchors. For that, we would need to store the list of the "largest" anchors for each slug (`["plants-4", "doors-4", "potatoes"]`). Maybe in page meta...?

Reply:
I think think that this is required or even desired. Because normally you want to avoid duplicated headings. It makes no sense in most cases and might be bad for SEO anyways.

Now imagine I am currently scribbling my post together. I duplicate blocks to try things out, and I remove them afterwards. I end up having id anchor of `anchor-2` etc everywhere. But in the end I just want to have a "just working" TOC and a clean link for my users. Thatswhy I would prefer to always fall back to normal headings.

Another note:
What if I have `heading` and `heading-1` and I delete `heading`. Will `heading-1` become `heading` afterwards. My 2cents are, that it should. Correct me if I am wrong with that idea.

Reply:
@NicoHood you are right, it does make sense for the writing and drafting. What I am worried about is the post-publish phase once the URLs are out there in the wild. Permalinks, cool urls don't change, that kind of stuff.

Reply:
@adamziel That is correct, but it is not an url, it is an anchor. Meaning the website will still function, but not scroll.

Lets say you have a heading called: "Red cars" which you rename to "Blue bike". Then the anchor (aka "url") will be "broken" too.

If someone wants a fixed anchor, he should enter his own id from the block settings. I am not sure if that is still possible with this PR, but that's what I'd expect the editor to do.

BTW: That is also how all other CMS work, Github and the Arch Linux Wiki as well. Just be careful in renaming the headings if you are refering to their autogenerated anchor. As easy as it is. I guess there is no "better" or perfect solution.

Reply:
There is no perfect solution here... There are pros and cons in all cases.
I don't think the `-1` etc suffixes will be used that often... If they do get used often it's indicative of a messy page structure, since having multiple headings with the same text is sub-optimal for screen-readers. The suffixes are there to avoid issues, but they are an edge-case and not the norm.
Manually-edited anchors are persistent, so if after publishing a page an author edits a page and changes the structure of headings, they _can_ manually edit the anchors and make them whatever they want.